File: go/market/pov_service.go
// SPDX-License-Identifier: MIT
//
// Proof-of-Vicinity (PoV) anti-Sybil trust service (single-node, in-memory). Go port of
// AetherNet.Market.IPoVService / InMemoryPoVService. Two users meet physically; their devices exchange
// a signed token over a short-range transport (BLE/NFC/NearLink). Over time a directed trust graph maps
// how many distinct humans have verified a profile.
//
// Signatures are REAL Ed25519 (crypto/ed25519, RFC 8032 — byte-identical to the C# Ed25519SigningService)
// over the canonical token body (BuildSignableTokenData = "SubjectUhid + TimestampTicks + Transport").
// The single-node service holds one identity key and produces both the witness and subject signatures
// with it; the two-party mesh exchange (each side counter-signs with its own key) is PoVTokenExchangeService.
//
// SEPARATION: the resulting PoVScore is a purely local anti-Sybil routing/identity signal — it attaches
// NO value semantics and never touches any money/reward layer.
package market

import (
	"crypto/ed25519"
	"crypto/rand"
	"sync"
	"time"
)

// PoVService is the Proof-of-Vicinity trust service.
type PoVService interface {
	// IssueToken issues a PoV token to subjectUhid (both signatures from this node's identity key).
	IssueToken(witnessUhid, subjectUhid string, transport PoVTransportType) (*PoVToken, error)
	// AcceptToken records an incoming token iff it cryptographically verifies.
	AcceptToken(token *PoVToken) error
	// GetScore returns the current PoV score for a UHID.
	GetScore(uhid string) PoVScore
	// VerifyToken reports whether the token is structurally and cryptographically valid.
	VerifyToken(token *PoVToken) bool
	// ReportDefection reduces the witness's weighted score by 20%.
	ReportDefection(witnessUhid, defectorUhid string)
}

// InMemoryPoVService is a single-node, in-memory PoVService for testing / single-node scenarios.
type InMemoryPoVService struct {
	mu              sync.Mutex
	tokensBySubject map[string][]*PoVToken // SubjectUhid -> tokens vouching for it
	scoreOverrides  map[string]float64     // WitnessUhid -> overridden score (post-defection)

	privateKey ed25519.PrivateKey
	publicKey  ed25519.PublicKey

	// OnTokenReceived fires when a token is issued or accepted.
	OnTokenReceived func(*PoVToken)
}

// NewInMemoryPoVService constructs a service with a fresh self-contained Ed25519 identity.
func NewInMemoryPoVService() (*InMemoryPoVService, error) {
	pub, priv, err := ed25519.GenerateKey(rand.Reader)
	if err != nil {
		return nil, err
	}
	return &InMemoryPoVService{
		tokensBySubject: make(map[string][]*PoVToken),
		scoreOverrides:  make(map[string]float64),
		privateKey:      priv,
		publicKey:       pub,
	}, nil
}

// IssueToken implements PoVService.
func (s *InMemoryPoVService) IssueToken(witnessUhid, subjectUhid string, transport PoVTransportType) (*PoVToken, error) {
	timestampTicks := TimeToTicks(time.Now().UTC())
	signable := BuildSignableTokenData(subjectUhid, timestampTicks, transport)
	// REAL Ed25519 over the canonical body; both signatures from this node's one key (single-node model).
	sig := ed25519.Sign(s.privateKey, signable)

	token := &PoVToken{
		WitnessUhid:      witnessUhid,
		SubjectUhid:      subjectUhid,
		TimestampTicks:   timestampTicks,
		TransportUsed:    transport,
		WitnessSignature: append([]byte(nil), sig...),
		SubjectSignature: append([]byte(nil), sig...),
	}
	if cb := s.OnTokenReceived; cb != nil {
		cb(token)
	}
	return token, nil
}

// AcceptToken implements PoVService — records only a token that cryptographically verifies.
func (s *InMemoryPoVService) AcceptToken(token *PoVToken) error {
	if !s.VerifyToken(token) {
		return nil // silently ignored, mirroring the C# reference
	}
	s.mu.Lock()
	s.tokensBySubject[token.SubjectUhid] = append(s.tokensBySubject[token.SubjectUhid], token)
	s.mu.Unlock()
	if cb := s.OnTokenReceived; cb != nil {
		cb(token)
	}
	return nil
}

// GetScore implements PoVService.
func (s *InMemoryPoVService) GetScore(uhid string) PoVScore {
	s.mu.Lock()
	list := s.tokensBySubject[uhid]
	tokens := make([]*PoVToken, len(list))
	copy(tokens, list)
	override, hasOverride := s.scoreOverrides[uhid]
	s.mu.Unlock()

	if len(tokens) == 0 {
		// A UHID with no inbound tokens still surfaces a stored defection override (defection penalises
		// witness UHIDs, which may not themselves be subjects).
		o := 0.0
		if hasOverride {
			o = override
		}
		return PoVScore{Uhid: uhid, UniqueWitnesses: 0, WeightedScore: o, LastUpdated: time.Now().UTC()}
	}

	witnesses := make(map[string]struct{}, len(tokens))
	for _, t := range tokens {
		witnesses[t.WitnessUhid] = struct{}{}
	}
	unique := len(witnesses)

	// Sigmoid-ish: w / (w + 1).
	score := float64(unique) / (float64(unique) + 1.0)
	if hasOverride {
		score = override
	}
	return PoVScore{Uhid: uhid, UniqueWitnesses: unique, WeightedScore: score, LastUpdated: time.Now().UTC()}
}

// VerifyToken implements PoVService — both signatures must be valid Ed25519 over the canonical body,
// the parties present and distinct.
func (s *InMemoryPoVService) VerifyToken(token *PoVToken) bool {
	if token == nil {
		return false
	}
	// Structural: both parties signed, both UHIDs present, and distinct.
	if len(token.WitnessSignature) == 0 || len(token.SubjectSignature) == 0 ||
		token.WitnessUhid == "" || token.SubjectUhid == "" ||
		token.WitnessUhid == token.SubjectUhid {
		return false
	}
	// Cryptographic: BOTH signatures valid over the canonical body.
	signable := token.SignableData()
	witnessValid := ed25519.Verify(s.publicKey, signable, token.WitnessSignature)
	subjectValid := ed25519.Verify(s.publicKey, signable, token.SubjectSignature)
	return witnessValid && subjectValid
}

// ReportDefection implements PoVService — reduces the witness's weighted score by 20%.
func (s *InMemoryPoVService) ReportDefection(witnessUhid, defectorUhid string) {
	score := s.GetScore(witnessUhid)
	penalised := score.WeightedScore * 0.8
	s.mu.Lock()
	s.scoreOverrides[witnessUhid] = penalised
	s.mu.Unlock()
}

File: typescript/src/market/PoVService.ts
// SPDX-License-Identifier: MIT
//
// Proof-of-Vicinity (PoV) anti-Sybil trust service (single-node, in-memory).
// TypeScript port of AetherNet.Market.IPoVService / InMemoryPoVService. Two users
// meet physically; their devices exchange a signed token over a short-range
// transport (BLE/NFC/NearLink). Over time a directed trust graph maps how many
// distinct humans have verified a profile.
//
// Signatures are REAL Ed25519 (Ed25519Service / TweetNaCl) over the canonical
// token body (buildSignableTokenData = "SubjectUhid + TimestampTicks + Transport").
// The single-node service holds one identity key and produces both the witness and
// subject signatures with it; the two-party mesh exchange (each side counter-signs
// with its own key) is PoVTokenExchangeService.
//
// SEPARATION: the resulting PoVScore is a purely local anti-Sybil routing/identity
// signal — it attaches NO value semantics and never touches any money/reward layer.

import { Ed25519Service } from "../security/Ed25519Service.js";
import {
  PoVToken,
  PoVTransportType,
  buildSignableTokenData,
  unixMsToTicks,
  type PoVScore,
} from "./PoVToken.js";

/** The Proof-of-Vicinity trust service. */
export interface IPoVService {
  issueToken(witnessUhid: string, subjectUhid: string, transport?: PoVTransportType): Promise<PoVToken>;
  acceptToken(token: PoVToken): Promise<void>;
  getScore(uhid: string): Promise<PoVScore>;
  verifyToken(token: PoVToken): Promise<boolean>;
  reportDefection(witnessUhid: string, defectorUhid: string): Promise<void>;
}

/** Single-node, in-memory {@link IPoVService} for testing / single-node scenarios. */
export class InMemoryPoVService implements IPoVService {
  private readonly tokensBySubject = new Map<string, PoVToken[]>();
  private readonly scoreOverrides = new Map<string, number>();
  private readonly privateKey: Uint8Array;
  private readonly publicKey: Uint8Array;

  /** Fires when a token is issued or accepted. */
  onTokenReceived?: (token: PoVToken) => void;

  constructor() {
    const kp = Ed25519Service.generateKeyPair();
    this.privateKey = kp.privateKey;
    this.publicKey = kp.publicKey;
  }

  async issueToken(
    witnessUhid: string,
    subjectUhid: string,
    transport: PoVTransportType = PoVTransportType.Ble,
  ): Promise<PoVToken> {
    const ticks = unixMsToTicks(BigInt(Date.now()));
    const signable = buildSignableTokenData(subjectUhid, ticks, transport);
    // REAL Ed25519 over the canonical body; both signatures from this node's one key (single-node model).
    const sig = Ed25519Service.sign(this.privateKey, signable);

    const token = new PoVToken({
      witnessUhid,
      subjectUhid,
      timestampTicks: ticks,
      transportUsed: transport,
      witnessSignature: sig,
      subjectSignature: sig,
    });
    this.onTokenReceived?.(token);
    return token;
  }

  async acceptToken(token: PoVToken): Promise<void> {
    // Record only a token that cryptographically verifies — both signatures valid + distinct parties.
    if (!(await this.verifyToken(token))) return;
    const list = this.tokensBySubject.get(token.subjectUhid) ?? [];
    list.push(token);
    this.tokensBySubject.set(token.subjectUhid, list);
    this.onTokenReceived?.(token);
  }

  async getScore(uhid: string): Promise<PoVScore> {
    const tokens = this.tokensBySubject.get(uhid) ?? [];
    const override = this.scoreOverrides.get(uhid);

    if (tokens.length === 0) {
      // A UHID with no inbound tokens still surfaces a stored defection override.
      return { uhid, uniqueWitnesses: 0, weightedScore: override ?? 0, lastUpdated: new Date() };
    }

    const unique = new Set(tokens.map((t) => t.witnessUhid)).size;
    // Sigmoid-ish: w / (w + 1).
    let score = unique / (unique + 1);
    if (override !== undefined) score = override;
    return { uhid, uniqueWitnesses: unique, weightedScore: score, lastUpdated: new Date() };
  }

  async verifyToken(token: PoVToken): Promise<boolean> {
    // Structural: both parties signed, both UHIDs present, and distinct.
    if (
      !token.witnessSignature?.length ||
      !token.subjectSignature?.length ||
      !token.witnessUhid ||
      !token.subjectUhid ||
      token.witnessUhid === token.subjectUhid
    ) {
      return false;
    }
    // Cryptographic: BOTH signatures valid over the canonical body.
    const signable = token.signableData();
    return (
      Ed25519Service.verify(this.publicKey, signable, token.witnessSignature) &&
      Ed25519Service.verify(this.publicKey, signable, token.subjectSignature)
    );
  }

  async reportDefection(witnessUhid: string, _defectorUhid: string): Promise<void> {
    const score = await this.getScore(witnessUhid);
    this.scoreOverrides.set(witnessUhid, score.weightedScore * 0.8);
  }
}
